src/app/GlobalAttributes.cpp

```diff
@@ -0,0 +1,132 @@
+ *    limitations under the License.
+ */
+#include <app/GlobalAttributes.h>
+#include <app/data-model-provider/MetadataLookup.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+
+bool IsSupportedGlobalAttributeNotInMetadata(AttributeId attributeId)
+{
+    for (auto & attr : GlobalAttributesNotInMetadata)
+    {
+        if (attr == attributeId)
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+DataModel::ActionReturnStatus ReadGlobalAttributeFromMetadata(DataModel::Provider * provider, const ConcreteAttributePath & path,
+                                                              AttributeValueEncoder & encoder)
+{
+    CHIP_ERROR err;
+
+    switch (path.mAttributeId)
+    {
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id: {
+        DataModel::ListBuilder<CommandId> builder;
+        err = provider->GeneratedCommands(path, builder);
+        if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+        auto buffer = builder.TakeBuffer();
+
+        return encoder.EncodeList([&buffer](const auto & encoder) {
+            for (auto id : buffer)
+            {
+                ReturnErrorOnFailure(encoder.Encode(id));
+            }
+            return CHIP_NO_ERROR;
+        });
+    }
+    case Clusters::Globals::Attributes::AcceptedCommandList::Id: {
+        DataModel::ListBuilder<DataModel::AcceptedCommandEntry> builder;
+        err = provider->AcceptedCommands(path, builder);
+        if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+        auto buffer = builder.TakeBuffer();
+
+        return encoder.EncodeList([&buffer](const auto & encoder) {
+            for (auto entry : buffer)
+            {
+                ReturnErrorOnFailure(encoder.Encode(entry.commandId));
+            }
+            return CHIP_NO_ERROR;
+        });
+    }
+    case Clusters::Globals::Attributes::AttributeList::Id: {
+        DataModel::ListBuilder<DataModel::AttributeEntry> builder;
+        err = provider->Attributes(path, builder);
+        if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+        auto buffer = builder.TakeBuffer();
+
+        return encoder.EncodeList([&buffer](const auto & encoder) {
+            bool addedExtraGlobals      = false;
+            constexpr auto lastGlobalId = GlobalAttributesNotInMetadata[ArraySize(GlobalAttributesNotInMetadata) - 1];
+            // If EventList is not supported. The GlobalAttributesNotInMetadata is missing one id here.
+            static_assert(lastGlobalId - GlobalAttributesNotInMetadata[0] == ArraySize(GlobalAttributesNotInMetadata),
+                          "Ids in GlobalAttributesNotInMetadata not consecutive (except EventList)");
+
+            for (auto entry : buffer)
+            {
+                if (!addedExtraGlobals && entry.attributeId > lastGlobalId)
```

Comment:
I am somewhat unclear why we sort these...this was taken from the existing implementation, but it does seem to make the code more complex. Is the spec requiring sorting?

Reply:
Spec does not require sorting.  ember attribute metadata _is_ sorted.

When implementing attributes not in metadata we kept the "act as if they are still in metadata" behavior.  But yes, we can probably nix this part and just put these things at the end, always.